advanced_source/cpp_autograd.rst

```diff
@@ -0,0 +1,442 @@
+
+   4.5000  4.5000
+   4.5000  4.5000
+  [ CPUFloatType{2,2} ]
+
+You should have got a matrix of ``4.5``. For explanations on how we arrive at this value,
+please see `the corresponding section in this tutorial <https://pytorch.org/tutorials/beginner/blitz/autograd_tutorial.html#gradients>`_.
+
+Now let's take a look at an example of vector-Jacobian product:
+
+.. code-block:: cpp
+
+  x = torch::randn(3, torch::requires_grad());
+  
+  y = x * 2;
+  while (y.norm().item<double>() < 1000) {
+    y = y * 2;
+  }
+    
+  std::cout << y << std::endl;
+  std::cout << y.grad_fn()->name() << std::endl;
+
+Out:
+
+.. code-block:: shell
+
+  -1021.4020
+    314.6695
+   -613.4944
+  [ CPUFloatType{3} ]
+  MulBackward1
+
+If we want the vector-Jacobian product, pass the vector to ``backward`` as argument:
+
+.. code-block:: cpp
+
+  auto v = torch::tensor({0.1, 1.0, 0.0001}, torch::kFloat);
+  y.backward(v);
+  
+  std::cout << x.grad() << std::endl;
+
+Out:
+
+.. code-block:: shell
+
+    102.4000
+   1024.0000
+      0.1024
+  [ CPUFloatType{3} ]
+
+You can also stop autograd from tracking history on Tensors with ``.requires_grad() == true``
+either by putting ``torch::NoGradGuard`` in a code block
+
+.. code-block:: cpp
+
+  std::cout << x.requires_grad() << std::endl;
+  std::cout << x.pow(2).requires_grad() << std::endl;
+  
+  {
+    torch::NoGradGuard no_grad;
+    std::cout << x.pow(2).requires_grad() << std::endl;
+  }
+
+
+Out:
+
+.. code-block:: shell
+
+  true
+  true
+  false
+
+Or by using ``.detach()`` to get a new Tensor with the same content but that does
+not require gradients:
+
+.. code-block:: cpp
+
+  std::cout << x.requires_grad() << std::endl;
+  y = x.detach();
+  std::cout << y.requires_grad() << std::endl;
+  std::cout << x.eq(y).all().item<bool>() << std::endl;
+
+Out:
+
+.. code-block:: shell
+
+  true
+  false
+  true
+
+For more information on C++ tensor autograd APIs such as ``grad`` / ``requires_grad`` /
+``is_leaf`` / ``backward`` / ``detach`` / ``detach_`` / ``register_hook`` / ``retain_grad``,
+please see `the corresponding C++ API docs <https://pytorch.org/cppdocs/api/classat_1_1_tensor.html>`_.
+
+Computing higher-order gradients in C++
+---------------------------------------
+
+One of the applications of higher-order gradients is calculating gradient penalty.
+Let's see an example of it using ``torch::autograd::grad``:
+
+.. code-block:: cpp
+
+  #include <torch/torch.h>
+  
+  auto model = torch::nn::Linear(4, 3);
+  
+  auto input = torch::randn({3, 4}).requires_grad_(true);
+  auto output = model(input);
+  
+  // Calculate loss
+  auto target = torch::randn({3, 3});
+  auto loss = torch::nn::MSELoss()(output, target);
+  
+  // Use norm of gradients as penalty
+  auto grad_output = torch::ones_like(output);
+  auto gradient = torch::autograd::grad({output}, {input}, /*grad_outputs=*/{grad_output},
+                  /*retain_graph=*/true, /*create_graph=*/true,
+                  /*allow_unused=*/true)[0];
```

Comment:
Why do you set allow_unused here?
Also retain_graph is already set to true if not specified and create_graph=true.